docs/docs/custom-components.mdx

```diff
@@ -0,0 +1,400 @@
+---
+id: custom-components
+sidebar_label: Custom Components & Policies
+title: Custom Components & Policies
+abstract: You can extend Rasa Open Source with custom NLU components and policies. This page provides a guide on how to develop your own custom components.
+---
+
+Rasa Open Source provides a variety of [NLU components](components) and
+[policies](policies) out of the box. You can customize these out of the box
+components or create your own components from scratch by using custom components.
+
+To use your custom component with Rasa Open Source it has to fulfill the following
+requirements:
+- It has to implement the [`GraphComponent` interface](#the-graphcomponent-interface)
+- It has to be [registered](#registering-components-with-the-model-configuration)
+  with the used model configuration
+- It has to be used in the [configuration file](#using-custom-components-in-your-model-configuration)
+- It has to use type annotations. Rasa Open Source makes use of the type annotations
+  to validate your model configuration.
+  [Forward references](https://www.python.org/dev/peps/pep-0484/#forward-references)
+  are not allowed. If you're using Python 3.7 you can use
+  [`from __future__ import annotations`](https://www.python.org/dev/peps/pep-0563/#enabling-the-future-behavior-in-python-3-7)
+  to get rid of forward references.
+
```

Comment:
Overview should include a bit more background on what is a `GraphComponent` conceptually (not the implementation details below) - since users making custom ones will want to understand what is this class they inherit from/why is it the same for both NLU & policies

Reply:
Could go in its own first section titled `Graph Components`

Reply:
done